### DISCUSSION TITLE
Feature request: Better TreeNode management

### DISCUSSION THREAD

**Discussion by VictorYing** [sparcians/map#547] [Ideas]
I'm thinking about working on a replacement for the `std::vector<std::unique_ptr<sparta::TreeNode>> to_delete_;` member inside of the `sparta::app::Simulation` base class. I don't like the existing mechanism because modelers who have requirements that the destructor of one TreeNode is called before the destructor of another TreeNode seem to be getting the habit of shuffling their TreeNodes into this vector into a particular order, and then relying on the fact that GNU libstdc++'s `std::~vector()` happens to destruct its elements from front to back. The example code in this repo already demonstrates this fragile pattern:
https://github.com/sparcians/map/blob/e736f0dabaa6ee4704ada11f63d6b36028c5d532/sparta/example/DynamicModelPipeline/src/ExampleSimulation.cpp#L693

I have two reasons not to like this pattern:
- I would prefer to have a default where nodes are destructed in the *reverse* order they are constructed. This is because I feel is natural to write code that constructs parent nodes before children, and to place some shared resource in the parent (e.g., a SpartaSharedPointerAllocator) that children may depend on (e.g., the children may hold shared pointer instances that need to get destructed before the allocator is destructed).
- The C++ language standards leave the order of destruction within most STL containers unspecified: https://stackoverflow.com/questions/6169125/order-of-destruction-of-elements-of-an-stdvector
   Models that work on GNU libstdc++ may break when linked against LLVM's libc++, where std::vector destructs elements back-to-front (consistent with the behavior the C++ standard imposes for raw arrays): https://godbolt.org/z/h3Pon4oT9

I suggest keeping `to_delete_;` (there's too much model code out there that uses it) while introducing a replacement API that registers TreeNodes to be managed by sparta::app::Simulation base class by storing them in some private member and ensuring the destructor destructs them in a well-specified order. Maybe in MAPv3 we can formally deprecate to_delete_ but I wonder if it's okay to go ahead and introduce a new solution now within the MAPv2 series.

**Comment by furuame** [comment]
I think it's reasonable to define the deallocation order of `to_delete_` to avoid unspecified behavior that you described. We could explicitly clean up `to_delete_` at Simulation's destructor.

On the other hand, I'm assuming you're building a custom model by DynamicModelPipeline. Defining another set of TreeNodes and the desired destruction behavior in your derived Simulator class might also work out. Just replace the strings :)

**Comment by kathlenemagnus-mips** [comment]
I have a related complaint that I'd like to share with you. Maybe you can come up with a suggestion on how to handle it.
```
class ParentUnit: public sparta::Unit
{
public:
    class ParentUnitParameterSet : public sparta::ParameterSet
    {
    public:
        ParentUnitParameterSet(sparta::TreeNode* n) : sparta::ParameterSet(n) {}
    }

    ParentUnit(sparta::TreeNode* node, const ParentUnitParameterSet* p)
    {
        tns_to_delete_.emplace_back(new sparta::ResourceTreeNode(
            node, "child_unit", sparta::TreeNode::GROUP_NAME_NONE,
            sparta::TreeNode::GROUP_IDX_NONE, "Child Unit", &child_rf_));

    }

private:
    sparta::ResourceFactory<ChildUnit, ChildUnit::ChildUnitParameterSet> child_rf_;
    std::vector<std::unique_ptr<sparta::ResourceTreeNode>> tns_to_delete_;
};
```
Sparta relies on the `ParentUnit` destructor to destruct the child nodes and the factories in the proper order. Declaring `tns_to_delete_` before `child_rf_` in the class definition will cause a memory error that can be very difficult to debug.

**Comment by klingaard** [comment]
I totally agree with everything you said.  I also agree and strongly suggest (if you haven't already) adding to discussion #509 that exact requirement.  In addition, I **have** come across implementations of `vector` the deletes in the reverse order.  Of course, it was an older Microsoft implementation (of course!) which I think they've changed since then.

With that being said, the `to_delete_` protected member variable in [Simulation.hpp](https://github.com/sparcians/map/blob/e736f0dabaa6ee4704ada11f63d6b36028c5d532/sparta/sparta/app/Simulation.hpp#L859) is flat-out superfluous.  In fact, you don't need to use it and can easily use your own local memory storage to maintain proper destruction order.

Now, onto @kathlenemagnus-mips rightful complaint.  :smile: 

The original concept of factory/tree node creation pairing was to use the following pattern in Simulation:
```
MySimulation::MySimulation(...)
{
      getResourceSet()->addResourceFactory<core_example::CPUFactory>();
}
```
In a nutshell, you register factories with the Simulation class and then use the protected member `to_delete_` later in simulation construction pulling that registered factory:
```
void MySimualtion::buildTree_()
{
    to_delete_(new sparta::ResourceTreeNode(getRoot(),
                                            "cpu",
                                            sparta::TreeNode::GROUP_NAME_NONE,
                                            sparta::TreeNode::GROUP_IDX_NONE,
                                            "CPU Node",
                                            getResourceSet()->getResourceFactory("cpu")));
}
```
Then, when simulation ends, alllll just works.  In theory.  :laughing: 

These are things I wouldn't mind changing for map_v3.

**Comment by VictorYing** [comment]
I ended up eliminating all usage of SpartaSharedPointerAllocator in my codebase, which led to my losing motivation for doing substantial work on this issue.  (We're still using SpartaSharedPointer, but just with simple `new`. The reason we abandoned Sparta's custom Allocator is for debuggability with AddressSanitizer & Valgrind. We saw very little performance impact in a model codebase whose runtime is dominated by other factors and when we were already linking against the fast malloc implementation like jemalloc.)

So I just wrote a comment in https://github.com/sparcians/map/discussions/509 as Knute suggested. I'm not sure what the timeline is for MapV3 and at what point it makes sense to start offering new well-defined behavior (probably under a new name, to avoid surprising behavior when users upgrade) before formally deprecating the old functionality.

**Reply by VictorYing** [reply]
If it's desired, I can volunteer to do the relatively simple job of just adding a new deletion mechanism now (to be maintained in parallel w/ old to_delete_, for at least a little while)

**Reply by klingaard** [reply]
Appreciate the offer for help.  As we discussed in the map_v3 discussion thread (#509) we'll just remove the `to_delete_` concept completely and force the modeler to use their own maintenance structures.

Regarding the deprecation of `SpartaSharedPointerAllocator`...

Before I added this class I too was using sanitizers and valgrind to find issues but grew frustrated with the general, "yes you have a memory problem" followed by "yeah, it's somewhere around here"  

The new Allocator was a nice "containment" class bounding the billions of small allocated objects to a handful of allowed allocations.  I could now catch the simulator allocating beyond what I expected, and when it did, I used the allocator's method [getOutstandingAllocatedObjects](https://sparcians.github.io/map/classsparta_1_1SpartaSharedPointerAllocator.html#a7bf4e642024aa42aeb06dd37d2e99697) to see which objects were sticking around beyond their life expectancy.  The class' intent was not to speed up simulation, but it didn't really hurt either.

**Reply by VictorYing** [reply]
To clarify, my primary issue was using a dangling reference to an object after it was destructed (and having CI to detect/prevent doing this). For example, this could happen with a dangling weak pointer. I know AddressSantiizer and Valgrind can be taught about custom allocators/pools, but it was easier to get rid of the custom allocator.

For leaks (memory growth), we appreciated and actively used getOustandingAllocatedObjects for a time. Today I'd consider other solutions to track live objects without changing memory allocation methods, such as maintaining a count or a map in constructors/destructors (and using mixin/CRTP inheritance to make this reusable).
